core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -2134,43 +2141,59 @@ object Stream {
       * }}}
       */
     def merge[O2 >: O](that: Stream[F, O2])(implicit F: Effect[F],
-                                            ec: ExecutionContext): Stream[F, O2] = {
-      def go(l: AsyncPull[F, Option[(Segment[O2, Unit], Stream[F, O2])]],
-             r: AsyncPull[F, Option[(Segment[O2, Unit], Stream[F, O2])]]): Pull[F, O2, Unit] =
-        l.race(r).pull.flatMap {
-          case Left(l) =>
-            l match {
-              case None =>
-                r.pull.flatMap {
-                  case None           => Pull.done
-                  case Some((hd, tl)) => Pull.output(hd) >> tl.pull.echo
-                }
-              case Some((hd, tl)) =>
-                Pull.output(hd) >> tl.pull.unconsAsync.flatMap(go(_, r))
-            }
-          case Right(r) =>
-            r match {
-              case None =>
-                l.pull.flatMap {
-                  case None           => Pull.done
-                  case Some((hd, tl)) => Pull.output(hd) >> tl.pull.echo
+                                            ec: ExecutionContext): Stream[F, O2] =
+      Stream.eval(async.semaphore(0)).flatMap { doneSem =>
+        Stream.eval(async.promise[F, Unit]).flatMap { interruptL =>
+          Stream.eval(async.promise[F, Unit]).flatMap { interruptR =>
+            Stream.eval(async.promise[F, Throwable]).flatMap { interruptY =>
+              Stream
+                .eval(async.unboundedQueue[F, Option[(F[Unit], Chunk[O2])]])
+                .flatMap { outQ => // note that the queue actually contains up to 2 max elements thanks to semaphores guarding each side.
+                  def runUpstream(s: Stream[F, O2], interrupt: Promise[F, Unit]): F[Unit] =
+                    async.semaphore(1).flatMap { quard =>
+                      s.chunks
```

Comment:
Why operate in chunks instead of segments?

Reply:
a) Segement cannot be tested for emptiness 
b) I want the cost of processing of the `Segment` to be done concurrently between s1 and s2 instead on resulting stream, essentially utilising more concurrency.